hibernate-core/src/main/java/org/hibernate/bytecode/internal/bytebuddy/HibernateMethodLookupDispatcher.java

```diff
@@ -142,25 +141,39 @@ public Void run() {
 			@SuppressWarnings({ "unchecked", "rawtypes" })
 			public Class<?> run() {
 				try {
+					Class<?>[] stackTrace;
 					if ( stackWalker != null ) {
-						Optional<Class<?>> clazzOptional = (Optional<Class<?>>) stackWalkerWalkMethod.invoke( stackWalker, new Function<Stream, Object>() {
+						stackTrace = (Class<?>[]) stackWalkerWalkMethod.invoke( stackWalker, new Function<Stream, Object>() {
 							@Override
 							public Object apply(Stream stream) {
 								return stream.map( STACK_FRAME_GET_DECLARING_CLASS_FUNCTION )
-										.skip( System.getSecurityManager() != null ? 6 : 5 )
-										.findFirst();
+										.limit( 16 )
+										.toArray( Class<?>[]::new );
 							}
 						});
-
-						if ( !clazzOptional.isPresent() ) {
-							throw new HibernateException( "Unable to determine the caller class" );
-						}
-
-						return clazzOptional.get();
 					}
 					else {
-						return SECURITY_ACTIONS.getCallerClass();
+						stackTrace = SECURITY_ACTIONS.getCallerClass();
 					}
+
+					// this shouldn't happen but let's be safe
+					if ( stackTrace.length < 4 ) {
+						return null;
+					}
+
+					boolean hibernateMethodLookupDispatcherDetected = false;
+					// start at the 4th frame and limit that to the 16 first frames
+					for ( int i = 3; i < Math.min( 16, stackTrace.length ); i++ ) {
+						if ( stackTrace[i].getName().equals( HibernateMethodLookupDispatcher.class.getName() ) ) {
+							hibernateMethodLookupDispatcherDetected = true;
+							continue;
+						}
+						if ( hibernateMethodLookupDispatcherDetected ) {
+							return stackTrace[i];
+						}
+					}
+
+					return null;
```

Comment:
Wouldn't the correct behavior be to throw an exception stating you couldn't determine the caller class? It looks like it was the previous behavior. Right now you're going to throw an exception saying something like "not authorized".

Reply:
Fixed.